GO: Various changes for 2.6.0

Snarling opened the pull request:
* Moved Go to top level scope object, instead of living on Player
* Custom save/load routine, to avoid huge data structures in save file and allow typesafety during loading. For an example of this see the differences in the FullSave test (which didn't even have board history included, which is the biggest savegame reduction)
* Lots of variable renaming and some reorganization, at least for the stuff that was previously in boardState/goConstants
* Game no longer stores history as full boards, now just saves the one previous board, as a simple board (no analytics).
* Less reassignment of Go.currentGame (previously Player.go.boardState), allowing equality of this variable to be used as a test for whether it is still the same game (e.g. after awaiting during async functions).
* Go.stats (previously Player.go.status) no longer contains all opponent entries from the getgo, now only contains entries when there is relevant data to store for the opponent.

There are some changes outside of go files as well. This is due to some changes I had to make to accommodate Jest, which included moving some netscript error handling functions to a new file, and I also renamed some of those functions at the same time.

This also replaces #1115 since the issues addressed there are also fixed here, and due to the scope of changes here there is conflict between the two.

Snarling committed:
Move GoRoot.tsx to Go/ui folder

Consistent with structure of other mechanics

Snarling committed:
Enum renaming

Renaming enum to singular+capitalized for standardization. Also added Go to the name because enums will be moved to be available in global enums file. Also capitalized some regular type names that were lowercase.

Snarling committed:
Some more changes

* Rename getPlayerStats to getOpponentStats (it's for getting the go stats for a particular opponent)
* Allow GoState.status to be a partial record instead of a full record so it can start empty. The individual entries are generated on demand if they are missing using the getOpponentStats function.
* getOpponentStats no longer regenerates an entire GoState if a particular item was missing.
* resetGoNodePower function only clears the stats for the opponents with actual entries, instead of using a hardcoded list of opponents.
* Fix some renamed variables that weren't updated in some files in previous commit
* Created an enums file and duplicated GoOpponent to exist there to be able to use the Enum helper for typechecking player input strings. Will expand on this next commit and remove enums from the goConstants file.
* Lint & format

Snarling committed:
More changes

* Prevent autopopulation of Player.go.status with all keys
* History and status pages now only show factions you have status entries for
* Removed hardcoded arrays of opponent names (no longer used)
* Allow SF14 to be set using dev menu (for testing)

ficocelliguy commented:
This looks good! One minor thought regarding your todos: We currently show the final state of the current subnet on the status page and main go board for the player to review if desired. I also show the previous game's final position on the history page. 

Sometimes games can end fairly abruptly, and I feel it is important for new players to be able to look at the final board state to better understand what worked (or to debug/improve their own ipvgo scripts). Having the prior game's endstate especially is helpful when determining if your script is doing well. These states can be kept in the save object in a simpler way, though, it doesn't need all of its analytics baked in like it currently has.

Snarling committed:
Test fixes

* Fix imports that weren't auto renamed
* Point enums imports to the Enums file, store more of the enums there
* Right now the enums are duplicated between the enums file and goConstants, will remove the goConstants ones in another commit.
* Removed a test that was no longer valid

Snarling commented:
Yeah I messed around a little bit trying to get the boardState to be nullable, but I ran into some issues and might just put that off for now and revisit at a later date. Mostly just didn't like that for a player who hasn't interacted with the mechanic at all during a run, it creates a board for them in their save file.

But on the topic of slimming down the save file, I have a couple thoughts I'll try out but I'm not sure if I will run into any problems here:

boardState.history: Board[] is pretty heavy when we might be talking about each board being possibly 13x13 and each point on the board being an object with a bunch of data itself.  An array of moves would make more sense (x, y, color) as a way to track history. But if there's a bunch of logic that relies on history being a Board[] right now, I might have trouble getting some of it to work with a format change.

boardState.board: I am thinking about not storing the actual PointState[][] but just storing something similar to the simplified board here, just showing ownership of points instead of all the other analytic data (chains, etc). If we store the board in a simplified state, is there anything else we need to store to faithfully recreate the board after loading the save, or can all the stuff be calculated afterwards (chains, liberties, etc)?

Snarling committed:
Fix weird enum bug

Enums not allowed to use "calculated values" like looking up a member from another enum, since this makes them add unwanted values to the runtime object. Switched to using strings directly and this fixed the unexpected runtime behavior.

Will probably add a test that ensures all opponents have an associated faction (other than none/WD)

Snarling commented:
dc7c2dc fixes a weird bug with unexpected TS behavior. It looks like when we reference one enum while defining another, TS automatically makes reverse lookup entries which is not wanted. It's assumed for string enums that all values are actually enum members, but this weird TS behavior broke that assumption. Here's GoOpponent before the fix:

![image](https://github.com/bitburner-official/bitburner-src/assets/84951833/25b1ccbf-55a2-4167-a4d9-3dd879e75afa)

Snarling committed:
Organize constants

Moved boardState/goConstants into three files in the main Go folder:
Constants.ts: contains constants (and a couple functions for now)
Types.ts: contains types
Enums.ts: contains enums
Then updated imports to reflect new locations

Snarling committed:
Fix test failures

There were some circular imports handled gracefully by webpack, but which caused jest to fail. The main cause was EnumHelper.ts importing NetscriptHelpers.tsx, which then in turn imports a whole bunch of other stuff. Specifically the entire UI was being imported and this is where it got tripped up by a circular import.

Fixed this by moving the specific functionality used by EnumHelper into separate files that don't require lots of imports themselves.

As a side effect, also renamed the very cumbersome helpers.makeRuntimeErrorMsg into just errorMessage

While I was trying to fix the import loop I also moved one of the go initializer funcitons out of the constants file and into boardState.ts

Snarling committed:
Move Go object off of player

Currently there is no save/load functionality, will add that in the next commit.

Snarling committed:
Prettier

Snarling committed:
Change history to a SimpleBoard

* History is now a single SimpleBoard instead of an array of Boards.
* Renamed GameState back to BoardState, but changed some confusing function names that misused terms Board vs. BoardState
* Still need to add back in Save/Load functionality. Next commit.

Snarling committed:
Add back save/load functionality

Save/load should now work (old data from earlier in the dev cycle will not be loaded since that data exists on Player)

Snarling commented:
@ficocelliguy I believe this should be fully functional again now if you want to look it over before I merge it. Otherwise I'll go ahead and merge it and wait a few days / a week before actually pushing the 2.6.0 release.

Also @d0sboots FYI in case you want to take a look. I also changed some stuff outside Go during this PR because it was causing issues after I moved Go to a top level object. Biggest non-go thing is I moved the home location of the helpers.makeRuntimeErrorMsg function: now it's called errorMessage and it's still available on helpers but actually lives in a different file.

d0sboots commented on test/jest/__snapshots__/FullSave.test.ts.snap:
```diff
@@ -548,6 +548,29 @@ exports[`Check Save File Continuity FactionsSave continuity 1`] = `
 }
 `;
 
+exports[`Check Save File Continuity GoSave continuity 1`] = `
```
Optional: Maybe have something other than a *completely* fresh game to test more of the save/load surface?

d0sboots reviewed (commented):
Change *all* the files! XD

ficocelliguy committed:
IPVGO: add previousPlayer as a dependency to useMemo so it correctly updates when the board is updated (#1124)

Snarling committed:
Destructure boardstate to get rid of eslint ignore statement

Still gross, still looking at improvements.

Snarling committed:
Fix issue with repeated moves not being calculated correctly

Evaluation boards were updating their own previousBoard, which was unnecessary. This led to the repeated move check actually checking the evaluated board against the current board, not the previous board.

Snarling committed:
getAllChains takes in a Board not a BoardState

Snarling committed:
findNeighbors takes in a Board not a BoardState

Snarling committed:
findAdjacentLibertiesForPoint takes in a Board not a BoardState

Snarling committed:
findAdjacentLibertiesAndAlliesForPoint takes in a Board not a BoardState

Snarling committed:
Change param from BoardState to Board for more functions

getAllNeighbors
getPlayerNeighbors
getAllNeighboringChains
findAdjacentPointsInChain
findLibertiesForChain
getAllPotentialEyes
findLibertiesForChain
checkTerritoryOwnership
getTerritoryScores
getDisputedTerritoryMoves
getExpansionMoveArray (also removed an unused param)

Snarling committed:
Update getBoardCopy function

No impact, it was previously unused.

Snarling committed:
Lots of stuff

Changed a whole bunch more functions to take in a Board instead of BoardState.
Changed functions that mutate an existing board (e.g. updateChains, updateCaptures) to no longer return that object (it's modified in place)
Parameter changes made the wonky destructuring for useMemo no longer work, since previousPlayer was not being requested by that function anymore.

Dealt with the useMemo thing by doing the following:
Removed the automatic rerender on GameBoard/GameBoardWrapper every 400ms, and removed the memoization after verifying rerenders only occurred when a turn has actually been taken via the UI.
Added an EventEmitter so that rerender signals can be sent when netscript functions cause changes on the board. Any successful cheats, ai moves, regular player moves, and board resets that occur via netscript while the player is on the UI should trigger rerenders.

Snarling commented:
After merging #1124, I messed around with some stuff, mostly to get rid of the eslint overrides.

First I tried using destructuring (95466d9), and this worked but it looked really stupid. This was just a fake way of making it seem like these functions were actually relying on e.g. previousPlayer.

I modified a bunch of the function signatures so they indicated whether they needed a whole boardState or just the board. Turns out the relevant functions just needed the board.

I removed the periodic rerender and found that rerenders were only being triggered when a move was completed. But with the periodic rerender missing, netscript interactions with Go would not trigger rerenders, so I added an EventEmitter and made netscript trigger that event when manipulating the board. Now rerendering is only performed when the board is actually updated (which requires running these calculations that were previously memoized), so there was no need for memoization and I removed the useMemos.

Snarling committed:
Remove testing flags

Removed a temporary global assignment for the Go object
commented out a bunch of console.debug lines

Snarling commented:
Merging this, followup changes can be done via another PR

Snarling committed:
Add some savegame conversion to keep go stats

This section can probably be removed at some point, it's just for people who had a game in the 2.6.0 dev cycle with go savedata in the old format.
